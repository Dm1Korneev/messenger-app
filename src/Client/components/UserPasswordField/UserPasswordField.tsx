import Visibility from '@mui/icons-material/Visibility';
import VisibilityOff from '@mui/icons-material/VisibilityOff';
import IconButton from '@mui/material/IconButton';
import InputAdornment from '@mui/material/InputAdornment';
import { TextField } from 'formik-material-ui';
import { useState } from 'react';

const ENTER_KEY_CODE = 13;

type Props = Omit<
  React.ComponentPropsWithoutRef<typeof TextField>,
  'margin' | 'label' | 'fullWidth' | 'type' | 'autoComplete' | 'InputProps'>

export const UserPasswordField = (props: Props) => {
  const [showPassword, setShowPassword] = useState(false);

  const handleOnMouseDownShowPassword = () => {
    setShowPassword(true);
  };

  const handleOnMouseUpShowPassword = () => {
    setShowPassword(false);
  };

  const handleOnKeyDownShowPassword = ({ keyCode }: React.KeyboardEvent<HTMLButtonElement>) => {
    if (keyCode === ENTER_KEY_CODE) {
      setShowPassword(true);
    }
  };

  const handleOnMouseOutShowPassword = () => {
    setShowPassword(false);
  };

  return (
    <TextField
      autoComplete="current-password"
      fullWidth
      InputProps={{
        endAdornment: (
          <InputAdornment position="end">
            <IconButton
              aria-label="Toggle password visibility"
              onBlur={handleOnMouseOutShowPassword}
              onKeyDown={handleOnKeyDownShowPassword}
              onKeyUp={handleOnMouseOutShowPassword}
              onMouseDown={handleOnMouseDownShowPassword}
              onMouseOut={handleOnMouseOutShowPassword}
              onMouseUp={handleOnMouseUpShowPassword}
              size="large"
            >
              {showPassword ? <Visibility /> : <VisibilityOff />}
            </IconButton>
          </InputAdornment>
        ),
      }}
      label="Password *"
      margin="normal"
      type={showPassword ? 'text' : 'password'}
      {...props}
    />
  );
};
