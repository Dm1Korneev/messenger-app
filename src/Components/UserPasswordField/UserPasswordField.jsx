import React from 'react';
import { TextValidator } from 'react-material-ui-form-validator';
import PropTypes from 'prop-types';
import InputAdornment from '@material-ui/core/InputAdornment';
import IconButton from '@material-ui/core/IconButton';
import Visibility from '@material-ui/icons/Visibility';
import VisibilityOff from '@material-ui/icons/VisibilityOff';

class UserPasswordField extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      showPassword: false,
    };
  }

  handleOnMouseDownShowPassword = () => {
    this.setState({ showPassword: true });
  };

  handleOnMouseUpShowPassword = () => {
    this.setState({ showPassword: false });
  };

  handleOnKeyDownShowPassword = ({ keyCode }) => {
    if (keyCode === 13) { // Enter key
      this.setState({ showPassword: true });
    }
  };

  handleOnMouseOutShowPassword = () => {
    this.setState({ showPassword: false });
  };

  render() {
    const { value, onChange } = this.props;
    const { showPassword } = this.state;

    return (
      <TextValidator
        margin="normal"
        label="Password *"
        fullWidth
        onChange={onChange}
        name="password"
        type={showPassword ? 'text' : 'password'}
        id="password"
        autoComplete="current-password"
        color="primary"
        value={value}
        validators={['required', 'minStringLength:6']}
        errorMessages={['this field is required', 'minimum length 6 symbols']}
        InputProps={{
          endAdornment: (
            <InputAdornment position="end">
              <IconButton
                aria-label="Toggle password visibility"
                onMouseDown={this.handleOnMouseDownShowPassword}
                onMouseUp={this.handleOnMouseUpShowPassword}
                onKeyDown={this.handleOnKeyDownShowPassword}
                onMouseOut={this.handleOnMouseOutShowPassword}
                onBlur={this.handleOnMouseOutShowPassword}
                onKeyUp={this.handleOnMouseOutShowPassword}
              >
                {showPassword ? <Visibility /> : <VisibilityOff />}
              </IconButton>
            </InputAdornment>
          ),
        }}
      />
    );
  }
}

UserPasswordField.defaultProps = {
  value: undefined,
};
UserPasswordField.propTypes = {
  value: PropTypes.string,
  onChange: PropTypes.func.isRequired,
};

export default UserPasswordField;

