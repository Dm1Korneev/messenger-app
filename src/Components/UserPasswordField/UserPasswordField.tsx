import React, { useState } from 'react';
import { TextField } from 'formik-material-ui';
import InputAdornment from '@material-ui/core/InputAdornment';
import IconButton from '@material-ui/core/IconButton';
import Visibility from '@material-ui/icons/Visibility';
import VisibilityOff from '@material-ui/icons/VisibilityOff';

const ENTER_KEY_CODE = 13;

type Props = Omit<
  React.ComponentPropsWithoutRef<typeof TextField>,
  'margin' | 'label' | 'fullWidth' | 'type' | 'autoComplete' | 'InputProps'>

export const UserPasswordField = (props: Props) => {
  const [showPassword, setShowPassword] = useState(false);

  const handleOnMouseDownShowPassword = () => {
    setShowPassword(true);
  };

  const handleOnMouseUpShowPassword = () => {
    setShowPassword(false);
  };

  const handleOnKeyDownShowPassword = ({ keyCode }: React.KeyboardEvent<HTMLButtonElement>) => {
    if (keyCode === ENTER_KEY_CODE) {
      setShowPassword(true);
    }
  };

  const handleOnMouseOutShowPassword = () => {
    setShowPassword(false);
  };

  return (
    <TextField
      margin="normal"
      label="Password *"
      fullWidth
      type={showPassword ? 'text' : 'password'}
      autoComplete="current-password"
      InputProps={{
        endAdornment: (
          <InputAdornment position="end">
            <IconButton
              aria-label="Toggle password visibility"
              onMouseDown={handleOnMouseDownShowPassword}
              onMouseUp={handleOnMouseUpShowPassword}
              onKeyDown={handleOnKeyDownShowPassword}
              onMouseOut={handleOnMouseOutShowPassword}
              onBlur={handleOnMouseOutShowPassword}
              onKeyUp={handleOnMouseOutShowPassword}
            >
              {showPassword ? <Visibility /> : <VisibilityOff />}
            </IconButton>
          </InputAdornment>
        ),
      }}
      {...props}
    />
  );
};
