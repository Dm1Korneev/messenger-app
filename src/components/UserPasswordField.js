import React from "react";
import { TextValidator } from "react-material-ui-form-validator";
import PropTypes from "prop-types";

// material-ui
import withStyles from "@material-ui/core/styles/withStyles";
import InputAdornment from "@material-ui/core/InputAdornment";
import Visibility from "@material-ui/icons/Visibility";
import VisibilityOff from "@material-ui/icons/VisibilityOff";
import IconButton from "@material-ui/core/IconButton";

class UserPasswordField extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      showPassword: false
    };
  }

  handleOnMouseDownShowPassword = () => {
    this.setState({ showPassword: true });
  };

  handleOnMouseUpShowPassword = () => {
    this.setState({ showPassword: false });
  };

  handleOnMouseOutShowPassword = () => {
    this.setState({ showPassword: false });
  };

  render() {
    const { value, onChange } = this.props;
    const { showPassword } = this.state;

    return (
      <TextValidator
        margin="normal"
        label="Password *"
        fullWidth
        onChange={onChange}
        name="password"
        type={showPassword ? "text" : "password"}
        id="password"
        autoComplete="current-password"
        color="primary"
        value={value}
        validators={["required", "minStringLength:6"]}
        errorMessages={["this field is required", "minimum length 6 symbols"]}
        InputProps={{
          endAdornment: (
            <InputAdornment position="end">
              <IconButton
                aria-label="Toggle password visibility"
                onMouseDown={this.handleOnMouseDownShowPassword}
                onMouseUp={this.handleOnMouseUpShowPassword}
                onMouseOut={this.handleOnMouseOutShowPassword}
              >
                {showPassword ? <Visibility /> : <VisibilityOff />}
              </IconButton>
            </InputAdornment>
          )
        }}
      />
    );
  }
}

UserPasswordField.propTypes = {
  value: PropTypes.string,
  onChange: PropTypes.func.isRequired
};

const styles = theme => ({});

export default withStyles(styles)(UserPasswordField);
