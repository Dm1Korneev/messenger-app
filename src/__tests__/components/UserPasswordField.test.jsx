import React from 'react';

import UserPasswordField from 'Components/UserPasswordField';

jest.mock('react-material-ui-form-validator', () => ({
  TextValidator: global.mockComponent('TextValidator'),
}));
jest.mock('@material-ui/core/InputAdornment', () => global.mockComponent('InputAdornment'));
jest.mock('@material-ui/icons/Visibility', () => global.mockComponent('Visibility'));
jest.mock('@material-ui/icons/VisibilityOff', () => global.mockComponent('VisibilityOff'));
jest.mock('@material-ui/core/IconButton', () => global.mockComponent('IconButton'));

describe('render UserPasswordField component', () => {
  let wrapper;

  const props = {
    value: 'TEST',
    onChange: jest.fn(),
  };

  beforeAll(() => {
    wrapper = global.mount(
      <UserPasswordField {...props} />,
    );
  });

  test('component is render', () => {
    expect(wrapper.find('UserPasswordField').length).toBe(1);
  });

  test('TextValidator subcomponent is render', () => {
    wrapper.update();
    expect(wrapper.find({ originalcomponent: 'TextValidator' }).length).toBe(1);
  });
});

describe('snapshot-test UserPasswordField component', () => {
  test('Renders correct properties', () => {
    const props = {
      value: 'TEST',
      onChange: jest.fn(),
    };

    global.mountExpect(
      <UserPasswordField {...props} />,
    ).toMatchSnapshot();
  });
});
