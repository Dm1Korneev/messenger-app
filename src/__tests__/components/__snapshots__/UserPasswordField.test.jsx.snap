// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot-test UserPasswordField component Renders correct properties 1`] = `
<WithStyles(UserPasswordField)
  onChange={[MockFunction]}
  value="TEST"
>
  <UserPasswordField
    classes={Object {}}
    onChange={[MockFunction]}
    value="TEST"
  >
    <Component
      InputProps={
        Object {
          "endAdornment": <Unknown
            position="end"
          >
            <Unknown
              aria-label="Toggle password visibility"
              onBlur={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseOut={[Function]}
              onMouseUp={[Function]}
            >
              <Unknown />
            </Unknown>
          </Unknown>,
        }
      }
      autoComplete="current-password"
      color="primary"
      errorMessages={
        Array [
          "this field is required",
          "minimum length 6 symbols",
        ]
      }
      fullWidth={true}
      id="password"
      label="Password *"
      margin="normal"
      name="password"
      onChange={[MockFunction]}
      type="password"
      validators={
        Array [
          "required",
          "minStringLength:6",
        ]
      }
      value="TEST"
    >
      <div
        originalcomponent="TextValidator"
        testprops={
          Object {
            "InputProps": Object {
              "endAdornment": <Unknown
                position="end"
              >
                <Unknown
                  aria-label="Toggle password visibility"
                  onBlur={[Function]}
                  onKeyDown={[Function]}
                  onKeyUp={[Function]}
                  onMouseDown={[Function]}
                  onMouseOut={[Function]}
                  onMouseUp={[Function]}
                >
                  <Unknown />
                </Unknown>
              </Unknown>,
            },
            "autoComplete": "current-password",
            "color": "primary",
            "errorMessages": Array [
              "this field is required",
              "minimum length 6 symbols",
            ],
            "fullWidth": true,
            "id": "password",
            "label": "Password *",
            "margin": "normal",
            "name": "password",
            "onChange": [MockFunction],
            "type": "password",
            "validators": Array [
              "required",
              "minStringLength:6",
            ],
            "value": "TEST",
          }
        }
      />
    </Component>
  </UserPasswordField>
</WithStyles(UserPasswordField)>
`;
